artichoke-backend/src/extn/core/string/trampoline.rs

```diff
@@ -42,31 +44,324 @@ pub fn add(interp: &mut Artichoke, mut value: Value, mut other: Value) -> Result
 
     let mut concatenated = s.clone();
     // XXX: This call doesn't do a check to see if we'll exceed the max allocation
-    //    size and may panic.
+    //    size and may panic or abort.
     concatenated.extend_from_slice(to_append);
     super::String::alloc_value(concatenated, interp)
 }
 
 pub fn push(interp: &mut Artichoke, mut value: Value, mut other: Value) -> Result<Value, Error> {
+    if value.is_frozen(interp) {
+        let s = unsafe { super::String::unbox_from_value(&mut value, interp)? };
+        let message = "can't modify frozen String: "
+            .chars()
+            .chain(s.inspect())
+            .collect::<super::String>();
+        return Err(FrozenError::from(message.into_vec()).into());
+    }
+
     let mut s = unsafe { super::String::unbox_from_value(&mut value, interp)? };
+    if let Ok(int) = other.try_convert_into::<i64>(interp) {
+        return match s.encoding() {
+            Encoding::Utf8 => {
+                // Safety:
+                //
+                // The string is reboxed before any intervening operations on the
+                // interpreter.
+                // The string is reboxed without any intervening mruby allocations.
+                unsafe {
+                    let string_mut = s.as_inner_mut();
+                    // XXX: This call doesn't do a check to see if we'll exceed the max allocation
+                    //    size and may panic or abort.
+                    string_mut
+                        .try_push_codepoint(int)
+                        .map_err(|err| RangeError::from(err.message()))?;
+                    let s = s.take();
+                    super::String::box_into_value(s, value, interp)
+                }
+            }
+            Encoding::Ascii => {
+                let byte = u8::try_from(int).map_err(|_| RangeError::from(format!("{int} out of char range")))?;
+                // Safety:
+                //
+                // The string is reboxed before any intervening operations on the
+                // interpreter.
+                // The string is reboxed without any intervening mruby allocations.
+                unsafe {
+                    let string_mut = s.as_inner_mut();
+                    // XXX: This call doesn't do a check to see if we'll exceed the max allocation
+                    //    size and may panic or abort.
+                    string_mut.push_byte(byte);
+                    if !byte.is_ascii() {
+                        string_mut.set_encoding(Encoding::Binary);
```

Comment:
Does this really need the conversion?

```
irb(main):001:0> a = "hello"
=> "hello"
irb(main):002:0> i = 64
=> 64
irb(main):003:0> a << i
=> "hello@"
irb(main):004:0> a.encoding
=> #<Encoding:UTF-8>
irb(main):005:0> a.force_encoding(Encoding::ASCII)
=> "hello@"
irb(main):006:0> a.encoding
=> #<Encoding:US-ASCII>
irb(main):007:0> a << i
=> "hello@@"
irb(main):008:0> a.encoding
=> #<Encoding:US-ASCII>
```

It looks like it survives as ASCII afterwards

Reply:
😉 try this again with `s << 255`. That's what the `!byte.is_ascii()` checks for.

Reply:
hah oof :grimacing: